src/portable/nordic/nrf5x/dcd_nrf5x.c

```diff
@@ -101,7 +101,11 @@ static void edpt_dma_start(volatile uint32_t* reg_startep)
     else
     {
       // Otherwise simply block wait
-      while ( _dcd.dma_running )  { }
+      while ( _dcd.dma_running )  {
+        // Manually poll the handler if interrupts are disabled
+        if ( __get_PRIMASK() )
```

Comment:
Do you want to check the USB IRQ status here too? Otherwise other interrupts will cause dcd_int_handler to run.

Reply:
This states that it will only poll the USB ISR when all interrupts are disabled.  `__disable_irq()` works by setting `PRIMASK` to `1`.  This simply polls the ISR in a tight loop, regardless of whether the USB IRQ is enabled.

It doesn't care whether other interrupts are enabled, and it doesn't care if there's a pending interurpt or not.  It will still poll regardless.

Reply:
hmm, there shouldn't be issue with this block waiting, I add a bit of logging and don't see it ever wait or defer the dma

```
// helper to start DMA
static void edpt_dma_start(volatile uint32_t* reg_startep)
{

  // Only one dma can be active
  if ( _dcd.dma_running )
  {
    if (SCB->ICSR & SCB_ICSR_VECTACTIVE_Msk)
    {
        TU_LOG2("dma deferred\r\n");
      // If called within ISR, use usbd task to defer later
      usbd_defer_func( (osal_task_func_t) edpt_dma_start, (void*) reg_startep, true );
      return;
    }
    else
    {
        TU_LOG2("dma waiting\r\n");
      // Otherwise simply block wait
      while ( _dcd.dma_running )  { }
    }
  }

  _dcd.dma_running = true;
  TU_LOG2("dma started\r\n");

  (*reg_startep) = 1;
  __ISB(); __DSB();
}

// DMA is complete
static void edpt_dma_end(void)
{
  TU_ASSERT(_dcd.dma_running, );
  _dcd.dma_running = false;

  TU_LOG2("dma ended\r\n");
}
```

Reply:
The problem is that `_dcd.dma_running` is set by the interrupt handler, and if you disable interrupts then `_dcd.dma_running` effectively becomes `1`, meaning it will never exit.